### PR TITLE
Use `is_{some,ok}_and` more in the compiler

### PR DESCRIPTION
**Pull request by compiler-errors**
slightly more fluent-reading code

### CODE REVIEW

File: compiler/rustc_codegen_cranelift/build_system/utils.rs [hunk 1]
```diff
@@ -290,7 +290,7 @@ impl Drop for LogGroup {
 }
 
 pub(crate) fn maybe_incremental(cmd: &mut Command) {
-    if is_ci() || std::env::var("CARGO_BUILD_INCREMENTAL").map_or(false, |val| val == "false") {
+    if is_ci() || std::env::var("CARGO_BUILD_INCREMENTAL").is_some_and(|val| val == "false") {
```
Comment (compiler-errors) [inline compiler/rustc_codegen_cranelift/build_system/utils.rs:293]: I'll revert the clif change.

**Review by lqd — APPROVED** [review]

File: compiler/rustc_trait_selection/src/traits/coherence.rs [hunk 2]
```diff
@@ -362,14 +362,11 @@ fn impl_intersection_has_impossible_obligation<'a, 'cx, 'tcx>(
...
+            infcx.evaluate_obligation(obligation).is_ok_and(|result| !result.may_apply())
         } else {
             // We use `evaluate_root_obligation` to correctly track intercrate
             // ambiguity clauses. We cannot use this in the new solver.
-            selcx.evaluate_root_obligation(obligation).map_or(
-                false, // Overflow has occurred, and treat the obligation as possibly holding.
-                |result| !result.may_apply(),
-            )
+            selcx.evaluate_root_obligation(obligation).is_ok_and(|result| !result.may_apply())
         }
```
Comment (dtolnay) [inline compiler/rustc_trait_selection/src/traits/coherence.rs:370]: Is the original comment here worth keeping in some form?

```diff
  // We use `evaluate_root_obligation` to correctly track intercrate
  // ambiguity clauses. We cannot use this in the new solver.
+ //
+ // If evaluate_root_obligation returns an error, overflow has occurred
+ // and we treat the obligation as possibly holding.
  selcx.evaluate_root_obligation(obligation).is_ok_and(|result| !result.may_apply())
```

But I am not familiar with this code. Quite possible what the comment said is just obvious to someone familiar with this code, and not worth keeping.

If it's in fact something subtle, maybe a terse `is_ok_and` isn't a good thing here.

```rust
match selcx.evaluate_root_obligation(obligation) {
    Ok(result) => !result.may_apply(),
    // Overflow has occurred. Treat the obligation as possibly holding.
    Err(_overflow) => false,
}
```

**Review by dtolnay — CHANGES_REQUESTED** [review]

File: compiler/rustc_trait_selection/src/traits/coherence.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/coherence.rs:370]: So, I originally asked for this comment in https://github.com/rust-lang/rust/pull/112163#discussion_r1218683757.

I think it's pretty obvious for someone who knows how overflow is handled in the solver, which is why I removed it, but on second though, I can re-add it just for good measure.

Comment (dtolnay) [reply compiler/rustc_trait_selection/src/traits/coherence.rs:370]: Either way is okay with me. If you're the one who originally said "this needs a comment" and you've changed your mind, I'm good with no comment.

**Review by dtolnay — APPROVED** [review]

**Pull request merged**